### PR TITLE
mon: allow removal of tier of ec overwritable pool

### PR DESCRIPTION
**Pull request by batrick**
Fixes: http://tracker.ceph.com/issues/22754

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -11336,10 +11336,12 @@ bool OSDMonitor::_check_remove_tier(
   // Apply CephFS-specific checks
   const FSMap &pending_fsmap = mon->mdsmon()->get_pending();
   if (pending_fsmap.pool_in_use(base_pool_id)) {
-    if (base_pool->type != pg_pool_t::TYPE_REPLICATED) {
-      // If the underlying pool is erasure coded, we can't permit the
-      // removal of the replicated tier that CephFS relies on to access it
-      *ss << "pool '" << base_pool_name << "' is in use by CephFS via its tier";
+    if (base_pool->type == pg_pool_t::TYPE_ERASURE &&
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: nit, could avoid accessing the internals of `pool` by using `base_pool->is_erasure() && !base_pool->allows_ecoverwrites()`. this approach also matches with its counterpart in `FileSystemCommandHandler::_check_pool()` .

**Comment by jcsp** [comment]
Same comment as @tchaikov , otherwise looks good

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (batrick) [reply src/mon/OSDMonitor.cc]: Done, thanks!

**Comment by jecluis** [comment]
So, is it okay to remove the tier if the pool is EC and support EC overwrites? Mind sharing a bit of why one is acceptable and the other is not?

**Comment by jcsp** [comment]
CephFS clients can't handle non-overwrite-supporting EC pools -- if we add a pool like that, the cephfs client will try and do overwrites and they'll fail horribly.

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -11336,10 +11336,11 @@ bool OSDMonitor::_check_remove_tier(
...
   if (pending_fsmap.pool_in_use(base_pool_id)) {
-    if (base_pool->type != pg_pool_t::TYPE_REPLICATED) {
-      // If the underlying pool is erasure coded, we can't permit the
-      // removal of the replicated tier that CephFS relies on to access it
-      *ss << "pool '" << base_pool_name << "' is in use by CephFS via its tier";
+    if (base_pool->is_erasure() && !base_pool->allows_ecoverwrites()) {
+      // If the underlying pool is erasure coded and does not allow EC
+      // overwrites, we can't permit the removal of the replicated tier that
+      // CephFS relies on to access it
+      *ss << "pool '" << base_pool_name << "' does not allow EC overwrites and is in use by CephFS via its tier";
```
Comment (jecluis) [inline src/mon/OSDMonitor.cc]: According to my simple copy/paste, this line is way over 80 cols. Mind hard wrapping it, please?

**Comment by jecluis** [comment]
@jcsp gotcha. ty.

File: src/mon/OSDMonitor.cc (hunk 2 above)
Comment (batrick) [reply src/mon/OSDMonitor.cc]: Fixed!

**Pull request merged**